consensus/core/src/synchronizer.rs

```diff
@@ -36,12 +36,142 @@ const FETCH_FROM_PEERS_TIMEOUT: Duration = Duration::from_millis(4_000);
 
 const MAX_FETCH_BLOCKS_PER_REQUEST: usize = 200;
 
+const MAX_AUTHORITIES_TO_FETCH_PER_BLOCK: usize = 2;
+
+struct BlocksGuard {
+    map: Arc<InflightBlocksMap>,
+    block_refs: BTreeSet<BlockRef>,
+    peer: AuthorityIndex,
+}
+
+impl Drop for BlocksGuard {
+    fn drop(&mut self) {
+        self.map.unlock_blocks(self.block_refs.clone(), self.peer)
+    }
+}
+
+// Keeps a mapping between the missing blocks that have been instructed to be fectched and the authorities
+// that are currently fetching them.
+struct InflightBlocksMap {
+    inner: Mutex<HashMap<BlockRef, BTreeMap<AuthorityIndex, u32>>>,
+}
+
+impl InflightBlocksMap {
+    fn new() -> Arc<Self> {
+        Arc::new(Self {
+            inner: Mutex::new(HashMap::new()),
+        })
+    }
+
+    /// Locks the blocks to be fetched for the assigned `peer_index`. We want to avoid re-fetching the
+    /// missing blocks from too many authorities at the same time, thus we limit the concurrency
+    /// per block by attempting to lock per block. If a block is already fetched by the maximum allowed
+    /// number of authorities, then the block ref will not be included in the returned set. The method
+    /// returns all the block refs that have been successfully locked and allowed to be fetched.
+    /// If `skip_checks = true` then no checks are made against the max authorities to fetch per block or
+    /// same peer already fetching the same block. This is currently used only from the scheduler to
+    /// allow us attempt fetching the blocks concurrently.
+    fn lock_blocks(
+        self: &Arc<Self>,
+        context: &Context,
+        missing_block_refs: BTreeSet<BlockRef>,
+        peer: AuthorityIndex,
+        skip_checks: bool,
+    ) -> Option<BlocksGuard> {
+        let hostname = context.committee.authority(peer).hostname.clone();
+
+        let mut blocks = BTreeSet::new();
+        let mut inner = self.inner.lock();
+
+        for block_ref in missing_block_refs {
+            // check that the number of authorities that are already instructed to fetch the block is not
+            // higher than the allowed and the `peer_index` has not already been instructed to do that.
+            let authorities = inner.entry(block_ref).or_default();
+            if skip_checks
+                || (authorities.len() < MAX_AUTHORITIES_TO_FETCH_PER_BLOCK
+                    && authorities.get(&peer).is_none())
+            {
+                authorities.entry(peer).and_modify(|c| *c += 1).or_insert(1);
+                blocks.insert(block_ref);
+                context
+                    .metrics
+                    .node_metrics
+                    .fetched_blocks_additional_authority
```

Comment:
Why this is "additional" authorities? It seems every lock increments the corresponding peer.

Reply:
Ah that's a product of refactoring - I'll fix that. Good catch